File: clients/typescript/ctxd-client/src/operation.ts
/**
 * Capability operations.
 *
 * Mirrors `ctxd_cap::Operation` and the OpenAPI `operations[]` enum.
 * The wire serialization matches the daemon's snake_case names exactly.
 *
 * Modeled as a string-valued const enum object (rather than `enum`) so
 * the values are inlined at usage sites and the bundle stays tiny.
 */
export const Operation = {
  /** Read events under a subject. */
  Read: "read",
  /** Write (append) events. */
  Write: "write",
  /** List subject paths. */
  Subjects: "subjects",
  /** FTS / vector search. */
  Search: "search",
  /** Admin operations (mint tokens, manage peers). */
  Admin: "admin",
} as const;

/** A single capability operation as emitted on the wire. */
export type Operation = (typeof Operation)[keyof typeof Operation];

/** Type guard: is `value` a known wire-format operation string? */
export function isOperation(value: unknown): value is Operation {
  return (
    typeof value === "string" &&
    Object.values(Operation).includes(value as Operation)
  );
}

File: clients/python/ctxd-py/src/ctxd/_operation.py
"""Capability operation enum.

Mirrors ``ctxd_cap::Operation`` and the OpenAPI ``operations[]`` enum.
The wire serialization matches the daemon's snake_case names exactly.
"""

from __future__ import annotations

from enum import Enum


class Operation(str, Enum):
    """Operations a capability token can authorize.

    The string values are the wire-format names used in the JSON body of
    ``/v1/grant`` and the wire-protocol ``Grant`` verb. Subclassing
    ``str`` keeps the enum members JSON-serializable without a custom
    encoder.
    """

    READ = "read"
    """Read events under a subject."""

    WRITE = "write"
    """Write (append) events."""

    SUBJECTS = "subjects"
    """List subject paths."""

    SEARCH = "search"
    """FTS / vector search."""

    ADMIN = "admin"
    """Admin operations (mint tokens, manage peers)."""

    def as_wire_str(self) -> str:
        """Wire-format string for this operation."""
        return self.value

    def __str__(self) -> str:  # pragma: no cover - trivial
        return self.value
